wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUClient.java

```diff
@@ -260,34 +263,67 @@ public NodeList getFileData(String logicalname, Long beginrow, Integer numrows,
     }
 
     /**
-     * Use this function to retrieve file metadata such as column information
+     * Use this function to retrieve file metadata such as column information,
+     * for superfiles the metadata from the first subfile will be returned.
      * @param logicalname    - Logical filename.
      * @param clustername    - Optional - The cluster the logical filename is associated with.
-     * @return Array of DFUDataColumns
+     * @return ArrayList of DFUDataColumnInfo
      * @throws Exception
      */
-    public DFUDataColumn[] getFileMetaData(String logicalname, String clustername) throws Exception
+    public List<DFUDataColumnInfo> getFileMetaData(String logicalname, String clustername) throws Exception
     {
         WsDfuServiceSoapProxy proxy = getSoapProxy();
-        DFUGetFileMetaDataRequest req = new DFUGetFileMetaDataRequest();
-        DFUDataColumn[] cols = null;
-
-        req.setLogicalFileName(logicalname);
-
-        if (clustername != null)
-            req.setClusterName(clustername);
 
+        List<DFUDataColumnInfo> cols=new ArrayList<DFUDataColumnInfo>();
+        String eclrecord=null;
+        //getFileMetadata fails for superfiles; use first subfile to retrieve record structure if this is the case
+        //also retrieve ecl to extract extra information (xpath, maxlength) not in getFileMetadata
         try
         {
-            DFUGetFileMetaDataResponse resp = proxy.DFUGetFileMetaData(req);
-            if (resp == null)
-                return cols;
+            DFUInfoResponse details=this.getFileInfo(logicalname, clustername);
+            if (details != null && details.getFileDetail() != null) {
+                eclrecord=details.getFileDetail().getEcl();
+                    if (details.getFileDetail().getIsSuperfile()) {
+                        SuperfileListRequest sar=new SuperfileListRequest();
+                        sar.setSuperfile(logicalname);
+                        SuperfileListResponse sresp=this.getSoapProxy().superfileList(sar);
+                        if (sresp != null && sresp.getSubfiles() != null && sresp.getSubfiles().length>0) {
+                            logicalname=sresp.getSubfiles()[0];
+                        } else {
+                            throw new Exception(logicalname + " is a superfile with no subfiles, cannot determine file structure");
+                        }
+                }
+            }
+        }
+        catch (Exception e)
+        {
+            String msg="Error calling DFUInfo for " + logicalname + ":" + e.getMessage();
+            if (verbose)
```

Comment:
I don't think we need this check, 
 Utils.println(System.out, msg, true, verbose); <-- this will handle the if (Verbose)